File: web/src/types/dividends.ts
export interface DividendPayment {
  ex_date: string
  amount: number
}

export interface DividendSummary {
  ticker: string
  name: string
  currency: string
  current_price: number
  annual_dividend: number
  dividend_yield: number
  growth_rate_5y: number | null
  growth_rate_all: number | null
  payment_frequency: number
  history: DividendPayment[]
}

export interface DripRequest {
  ticker: string
  initial_shares: number
  years?: number
  dividend_growth_rate?: number | null
  price_growth_rate?: number
}

export interface DripPoint {
  year: number
  shares: number
  share_price: number
  annual_dividend_income: number
  annual_tax: number
  annual_dividend_after_tax: number
  portfolio_value: number
  total_dividends_received: number
  total_tax_paid: number
}

export interface DripProjection {
  ticker: string
  initial_shares: number
  initial_investment: number
  dividend_growth_rate: number
  price_growth_rate: number
  years: number
  points: DripPoint[]
}

export interface DripCompareTickerInput {
  ticker: string
  dividend_growth_rate?: number | null
  price_growth_rate: number
  share_price_override?: number | null
}

export interface DripCompareRequest {
  tickers: DripCompareTickerInput[]
  initial_shares: number
  years?: number
  annual_contribution?: number
  dividend_tax_rate?: number
}

export interface DripCompareItem {
  ticker: string
  name: string
  initial_share_price: number
  dividend_yield: number
  dividend_growth_rate: number
  price_growth_rate: number
  initial_investment: number
  ending_balance: number
  total_return_pct: number
  avg_annual_return_pct: number
  final_annual_income: number
  final_annual_income_after_tax: number
  total_dividends_paid: number
  total_tax_paid: number
  yield_on_cost: number
  points: DripPoint[]
  error?: string
}

export interface DripCompareResponse {
  initial_shares: number
  years: number
  results: DripCompareItem[]
}

export interface TickerSearchResult {
  symbol: string
  name: string
  exchange: string
  type: string
}

export interface HoldingRequest {
  ticker: string
  shares: number
}

export interface HoldingIncome {
  ticker: string
  name: string
  shares: number
  price?: number
  annual_dividend_per_share?: number
  annual_income?: number
  value?: number
  yield_pct?: number
  currency?: string
  error?: string
}

export interface PortfolioIncome {
  total_annual_income: number
  total_portfolio_value: number
  weighted_yield: number
  holdings: HoldingIncome[]
}

File: api/schemas/dividends.py
"""Pydantic schemas for the dividend calculator API."""

from typing import List, Optional

from pydantic import BaseModel


class DividendPaymentResponse(BaseModel):
    ex_date: str
    amount: float


class DividendSummaryResponse(BaseModel):
    ticker: str
    name: str
    currency: str
    current_price: float
    annual_dividend: float
    dividend_yield: float
    growth_rate_5y: Optional[float] = None
    growth_rate_all: Optional[float] = None
    payment_frequency: int
    history: List[DividendPaymentResponse]


class DripRequest(BaseModel):
    ticker: str
    initial_shares: float
    years: int = 20
    dividend_growth_rate: Optional[float] = None
    price_growth_rate: float = 7.0
    annual_contribution: float = 0.0


class DripPointResponse(BaseModel):
    year: int
    shares: float
    share_price: float
    annual_dividend_income: float
    annual_tax: float
    annual_dividend_after_tax: float
    portfolio_value: float
    total_dividends_received: float
    total_tax_paid: float


class DripProjectionResponse(BaseModel):
    ticker: str
    initial_shares: float
    initial_investment: float
    dividend_growth_rate: float
    price_growth_rate: float
    years: int
    points: List[DripPointResponse]


class DripCompareTickerInput(BaseModel):
    ticker: str
    dividend_growth_rate: Optional[float] = None  # None = use historical CAGR
    price_growth_rate: float = 7.0
    share_price_override: Optional[float] = None  # None = use current market price


class DripCompareRequest(BaseModel):
    tickers: List[DripCompareTickerInput]
    initial_shares: float
    years: int = 20
    annual_contribution: float = 0.0
    dividend_tax_rate: float = 0.0


class DripCompareItem(BaseModel):
    ticker: str
    name: str
    initial_share_price: float
    dividend_yield: float
    dividend_growth_rate: float
    price_growth_rate: float
    initial_investment: float
    # Summary stats
    ending_balance: float
    total_return_pct: float
    avg_annual_return_pct: float
    final_annual_income: float
    final_annual_income_after_tax: float
    total_dividends_paid: float
    total_tax_paid: float
    yield_on_cost: float
    points: List[DripPointResponse]
    error: Optional[str] = None


class DripCompareResponse(BaseModel):
    initial_shares: float
    years: int
    results: List[DripCompareItem]


class TickerSearchResultResponse(BaseModel):
    symbol: str
    name: str
    exchange: str
    type: str


class HoldingRequest(BaseModel):
    ticker: str
    shares: float


class PortfolioIncomeRequest(BaseModel):
    holdings: List[HoldingRequest]


class HoldingIncomeResponse(BaseModel):
    ticker: str
    name: str
    shares: float
    price: Optional[float] = None
    annual_dividend_per_share: Optional[float] = None
    annual_income: Optional[float] = None
    value: Optional[float] = None
    yield_pct: Optional[float] = None
    currency: Optional[str] = None
    error: Optional[str] = None


class PortfolioIncomeResponse(BaseModel):
    total_annual_income: float
    total_portfolio_value: float
    weighted_yield: float
    holdings: List[HoldingIncomeResponse]
